docs: add Form Layout doc text and TS examples (#239)
vaadin/docs: articles/ds/components/form-layout/index.asciidoc

= Form Layout

// tag::description[]
Form Layout allows you to build responsive forms with multiple columns and to position input labels on top or to the side of the input.
// end::description[]

[.example]
--

[source,typescript]
----
include::../../../../frontend/demo/component/formlayout/form-layout-basic.ts[render,tags=snippet,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/formlayout/FormLayoutBasic.java[render,tags=snippet,indent=0,group=Java]
// ----

--

.Work In Progress
[NOTE]
====
The usage examples for this component are currently in progress.
In the meantime, you can view the previous examples for the web component and the Java/Flow component.

[.buttons]
- https://vaadin.com/components/vaadin-form-layout/html-examples[Web Component Examples]
- https://vaadin.com/components/vaadin-form-layout/java-examples[Java Component Examples]
====

== Columns

> Form Layout sports two columns by default meaning it displays two input fields per line. On smaller viewports it adjusts to a single column layout.

Author: Done.

=== Custom Layout

You can define how many columns Form Layout should use based on the screen width.

[.example]
--

> [source,typescript]

Author: I disagree because of the TS code present on the example. I changed the first example to `html`, though.

A multi column layout is more prone to cause confusion and to be misinterpreted by the user.

However, closely related fields can be placed in line without issue.
For example, first and last name, address fields such as postal code and city, as well as ranged input for dates, time, currency, etc.

=== Colspan

When using a multi column layout you can define a colspan for each component.
Colspan determines how many columns a component extends or stretches across.

For example, if you have a Form Layout with three columns and a component’s colspan is set to three, it will take up the entire width of the Form Layout.

[.example]
--

[source,typescript]
----
include::../../../../frontend/demo/component/formlayout/form-layout-colspan.ts[render,tags=snippet,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/formlayout/FormLayoutColspan.java[render,tags=snippet,indent=0,group=Java]
// ----

--

== Label Position

Input fields’ built-in labels are positioned above the input.
Form Layout supports side positioned labels provided they are wrapped in Form Items and the label position is set to “aside”.

> The only reason for wrapping labels in Form Items is to put the labels to the side of the input.

Author: It was removed on purpose. See this [comment](https://github.com/vaadin/docs/pull/239#discussion_r589522535).

=== Top

Forms using top-positioned labels generally produce faster completion rates because they provide a consistent scanning pattern (top-down as opposed to a zig-zag) while minimizing the distance between the label and input.

Top-positioned labels are also less prone to cause layout issues due to variable label lengths as is the case of multilingual applications.

> They do however result in vertically longer forms which is why sectioning is important.

Author: Good catch. Done.

=== Side

Side-positioned labels help reduce a form’s overall height.
This is especially useful for longer forms and/or when vertical space is limited.
They’re also often used when there is a need to compare numeric data.

[.example]
--

[source,typescript]
----
include::../../../../frontend/demo/component/formlayout/form-layout-side.ts[render,tags=snippet,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/formlayout/FormLayoutSide.java[render,tags=snippet,indent=0,group=Java]
// ----

--

Aim for similar length labels to keep the distance between the labels and input fields consistent.
Inconsistent spacing can cause slower completion rates.

Forms using this position require more horizontal space which isn’t always ideal on narrow viewports.
It’s recommended to configure Form Layout to use top-positioned labels on smaller screens.

.Accessibility
[CAUTION]
====
Form Item suffers from two accessibility issues: required indicators cannot be displayed and screen readers are unable to announce a field’s label.
====

=== Responsive Label Position

> Like the number of columns, the label position is configurable based on the viewport’s width.

Author: Done. Thanks.

> For example, you can position the labels on the side when there’s ample horizontal space available, and on top on narrower screens.

Author: Yes. There's a comment on the document from Rolf to skip it for now.

== Native Input Fields

Form Item allows you to set a label for any type of component that you want to use in a Form Layout. It supports both Vaadin components and native HTML components.

[.example]
--

[source,typescript]
----
include::../../../../frontend/demo/component/formlayout/form-layout-native-input.ts[render,tags=snippet,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/formlayout/FormLayoutNativeInput.java[render,tags=snippet,indent=0,group=Java]
// ----

--

== Best Practises

=== Sectioning

Longer forms should be split into smaller, more manageable and user-friendly sections using subheadings, <<../tabs#,Tabs>>, <<../details#,Details>> or separate views when possible.

> Each section should consist of related content and/or fields.

Author: It was decided to be skipped by Rolf.

= Form Layout

// tag::description[]
Form Layout allows you to build responsive forms with multiple columns and to position input labels on top or to the side of the input.
// end::description[]

[.example]
--

[source,html]
----

> include::../../../../frontend/demo/component/formlayout/form-layout-basic.ts[render,tags=snippet,indent=0,group=TypeScript]

Reviewer: ```suggestion
include::{root}/frontend/demo/component/formlayout/form-layout-basic.ts[render,tags=snippet,indent=0,group=TypeScript]
```

----

// [source, java]
// ----

> // include::../../../../src/main/java/com/vaadin/demo/component/formlayout/FormLayoutBasic.java[render,tags=snippet,indent=0,group=Java]

Reviewer: ```suggestion
// include::{root}/src/main/java/com/vaadin/demo/component/formlayout/FormLayoutBasic.java[render,tags=snippet,indent=0,group=Java]
```

// ----

--

.Work In Progress
[NOTE]
====
The usage examples for this component are currently in progress.
In the meantime, you can view the previous examples for the web component and the Java/Flow component.

[.buttons]
- https://vaadin.com/components/vaadin-form-layout/html-examples[Web Component Examples]
- https://vaadin.com/components/vaadin-form-layout/java-examples[Java Component Examples]
====

== Columns

Form Layout sports two columns by default meaning it displays two input fields per line.
On smaller viewports it adjusts to a single column layout.

=== Custom Layout

You can define how many columns Form Layout should use based on the screen width.

[.example]
--

[source,typescript]
----

> include::../../../../frontend/demo/component/formlayout/form-layout-custom-layout.ts[render,tags=snippet,indent=0,group=TypeScript]

Reviewer: ```suggestion
include::{root}/frontend/demo/component/formlayout/form-layout-custom-layout.ts[render,tags=snippet,indent=0,group=TypeScript]
```

----

// [source, java]
// ----

> // include::../../../../src/main/java/com/vaadin/demo/component/formlayout/FormLayoutCustomLayout.java[render,tags=snippet,indent=0,group=Java]

Reviewer: ```suggestion
// include::{root}/src/main/java/com/vaadin/demo/component/formlayout/FormLayoutCustomLayout.java[render,tags=snippet,indent=0,group=Java]
```

// ----

--

> A single column layout is generally preferable to a multi column one.

Reviewer: ```suggestion
A single column layout is preferable to a multi column one.
```

A multi column layout is more prone to cause confusion and to be misinterpreted by the user.

However, closely related fields can be placed in line without issue.
For example, first and last name, address fields such as postal code and city, as well as ranged input for dates, time, currency, etc.

=== Colspan

When using a multi column layout you can define a colspan for each component.
Colspan determines how many columns a component extends or stretches across.

> For example, if you have a Form Layout with three columns and a component’s colspan is set to three, it will take up the entire width of the Form Layout.

Reviewer: ```suggestion
For example, if you have a Form Layout with three columns and a component’s colspan is set to three, it takes up the entire width of the Form Layout.
```

[.example]
--

[source,typescript]
----

> include::../../../../frontend/demo/component/formlayout/form-layout-colspan.ts[render,tags=snippet,indent=0,group=TypeScript]

Reviewer: ```suggestion
include::{root}/frontend/demo/component/formlayout/form-layout-colspan.ts[render,tags=snippet,indent=0,group=TypeScript]
```

----

// [source, java]
// ----

> // include::../../../../src/main/java/com/vaadin/demo/component/formlayout/FormLayoutColspan.java[render,tags=snippet,indent=0,group=Java]

Reviewer: ```suggestion
// include::{root}/src/main/java/com/vaadin/demo/component/formlayout/FormLayoutColspan.java[render,tags=snippet,indent=0,group=Java]
```

// ----

--

== Label Position

Input fields’ built-in labels are positioned above the input.
Form Layout supports side positioned labels provided they are wrapped in Form Items and the label position is set to “aside”.

The only reason for wrapping labels in Form Items is to put the labels to the side of the input.

=== Top

Forms using top-positioned labels generally produce faster completion rates because they provide a consistent scanning pattern (top-down as opposed to a zig-zag) while minimizing the distance between the label and input.

Top-positioned labels are also less prone to cause layout issues due to variable label lengths as is the case of multilingual applications.

They do however result in vertically longer forms which is why <<#sectioning,sectioning>> is important.

=== Side

Side-positioned labels help reduce a form’s overall height.
This is especially useful for longer forms and/or when vertical space is limited.
They’re also often used when there is a need to compare numeric data.

[.example]
--

[source,typescript]
----

> include::../../../../frontend/demo/component/formlayout/form-layout-side.ts[render,tags=snippet,indent=0,group=TypeScript]

Reviewer: ```suggestion
include::.{root}/frontend/demo/component/formlayout/form-layout-side.ts[render,tags=snippet,indent=0,group=TypeScript]
```

----

// [source, java]
// ----

> // include::../../../../src/main/java/com/vaadin/demo/component/formlayout/FormLayoutSide.java[render,tags=snippet,indent=0,group=Java]

Reviewer: ```suggestion
// include::{root}/src/main/java/com/vaadin/demo/component/formlayout/FormLayoutSide.java[render,tags=snippet,indent=0,group=Java]
```

// ----

--

Aim for similar length labels to keep the distance between the labels and input fields consistent.
Inconsistent spacing can cause slower completion rates.

Forms using this position require more horizontal space which isn’t always ideal on narrow viewports.
It’s recommended to configure Form Layout to use top-positioned labels on smaller screens.

.Accessibility
[CAUTION]

> ====
> Form Item suffers from two accessibility issues: required indicators cannot be displayed and screen readers are unable to announce a field’s label.
> ====

Reviewer: ```suggestion
Form Item suffers from two accessibility issues: required indicators cannot be displayed and screen readers are unable to announce a field’s label.
```
```suggestion
====
Form Item suffers from two accessibility issues: required indicators cannot be displayed and screen readers are unable to announce a field’s label.
====
```

=== Responsive Label Position

Like the number of columns, *the label position is configurable based on the viewport’s width*.
For example, you can position the labels on the side when there’s ample horizontal space available, and on top on narrower screens.

== Native Input Fields

> Form Item allows you to set a label for any type of component that you want to use in a Form Layout. It supports both Vaadin components and native HTML components.

Reviewer: ```suggestion
Form Item allows you to set a label for any type of component that you want to use in a Form Layout.
It supports both Vaadin components and native HTML components.
```

[.example]
--

[source,typescript]
----

> include::../../../../frontend/demo/component/formlayout/form-layout-native-input.ts[render,tags=snippet,indent=0,group=TypeScript]

Reviewer: ```suggestion
include::{root}/frontend/demo/component/formlayout/form-layout-native-input.ts[render,tags=snippet,indent=0,group=TypeScript]
```

----

// [source, java]
// ----

> // include::../../../../src/main/java/com/vaadin/demo/component/formlayout/FormLayoutNativeInput.java[render,tags=snippet,indent=0,group=Java]

Reviewer: ```suggestion
// include::{root}/src/main/java/com/vaadin/demo/component/formlayout/FormLayoutNativeInput.java[render,tags=snippet,indent=0,group=Java]
```

A multi column layout is more prone to cause confusion and to be misinterpreted by the user.

However, closely related fields can be placed in line without issue.
For example, first and last name, address fields such as postal code and city, as well as ranged input for dates, time, currency, etc.

=== Colspan

When using a multi column layout you can define a colspan for each component.
Colspan determines how many columns a component extends or stretches across.

For example, if you have a Form Layout with three columns and a component’s colspan is set to three, it will take up the entire width of the Form Layout.

[.example]
--

[source,typescript]
----
include::../../../../frontend/demo/component/formlayout/form-layout-colspan.ts[render,tags=snippet,indent=0,group=TypeScript]
----

// [source, java]
// ----
// include::../../../../src/main/java/com/vaadin/demo/component/formlayout/FormLayoutColspan.java[render,tags=snippet,indent=0,group=Java]
// ----

--

== Label Position

Input fields’ built-in labels are positioned above the input. Form Layout supports side positioned labels provided they are wrapped in Form Items and the label position is set to “aside”.

.NOTE

> [NOTE]

Author: I couldn't come up with another heading to the note.

Reviewer: I would remove the admonition. Let the text be a regular paragraph.

Author: Removed.
